### ISSUE TITLE
I need to learn more about developing scripts

### ISSUE THREAD

**Issue by EntityDev** [BLCM/BLCMods#574]
Has anyone created any in-depth documentation regarding the creation of hotfixes/scripts? If so, can someone please be so kind as to point them out?

I have searched exhaustively and was able to find only a very short discussion on the topic, and a short youtube video, neither of which go into great detail.

My main interest is in learning how to swap enemies. I have studied the only source example I am aware of, the Enemy Randomizer. Unfortunately, it is so voluminous that it is difficult to break it apart and make sense of the process.

A few basic questions, if I may:
1. I note in all scripts there is a pound sign (#) preceding all hotfix key and value definitions, whereas the 'set' statements are not preceded with one. What is the meaning of the pound sign? As a developer, my first instinct is that these are comments. But each of the commented lines have key/value pairs that are referenced in a later 'Set' statement. Are the pound signs comments or do they indicate that a line is a key/value pair definition, or maybe both?

2. I'd like to learn the syntax of creating the hotfixes/scripts. For example, a Set statement contains the Set call, a space, (apparently) the name of a function, a space, (apparently) a variable name, a space, then a (usually giant) string describing something or other to satisfay this last argument. Can someone please explain what the definition for each argument is? Is there some documentation I could study?
 
3. In the Community Patch file, the first node, "#<Original Gearbox hotfix data - Don't uncheck>". Do the hotfixes described in this section need to executed in conjunction with any/every script I create or find, and execute? I assume that if this is true, they are not permanent fixes, and are possibly run, invisibly, whenever I run the game. Can someone provide insight into these, please?

4. Often, when I 'obj dump' a class in the console, more lines are generated than can fit in the console window. Are the results of this dump written out to a text file? If so, what is the path? Is there a way to force the output to pause, so that I can view the results a page at a time in the console window? Similar to DOS 6.0 "|more" (pipe symbol). I tried many variations but was not successful.

4. How can I generate nice, clean lists of:
    a. Map names.
    b. Enemy mixes and single enemies.
    c. The integers associated with each member of a mix..

If there is a way to dump these things to disk I'd sure like to learn how.

5. Swapping - can someone please give me the basic steps to swap one single enemy for another? Or, perhaps, replacing a member of a mx with another enemy, or member of a different mix.

I am unsure what steps are the minimum requred to effectively swap one thing for another.

I note the first node appears to be loading maps into memory, though I can't be sure that is what this section is doing. It appears to define a list (LevelList), then populate it. It appears to define a main map (PersistenMap), then several SecondaryMaps. These lists of SecondaryMaps are huge, so I assume they're lists of maps which can be referenced in-game by the PersistentMap. Can someone shed light on this process, please? I'm hamstrung because I don't know the meaning of each argument.

The SecondaryMaps sections contain class names which do not appear to be maps, as I cannot 'Open" them. For example, I can open Fyrestone_P, but attempting to open Fyrestone_Combat takes me an area that is completely black. Attempting to open other "maps" such as tnose ending with "Audio" or "Dynamic" will crash the client. Obviously these other files are not main maps. My guess is they are support libraries for a given map.

Can someone go into detail about this, please? Is it even necessary to load any maps, if my only intention is to swap a enemy with another that is available in the map I go into?

6. How can I generate a list of all single enemies and enemy mixes that are available for a given map?

Thanks in advance for any help with the above.

**Comment by Ravenjim** [comment]
For all that you can either check the wiki we have here. Cody has made some excelent exampels or join us on ShadowEvils Discord server and ask around.

**Comment by EntityDev** [comment]
Thank you. I have read the wiki and, no offense to anyone, but I find it lacking. I'll have to pass on the Discord app - I have enough clutter already - but I do thank you for the invite.

I have put in some study time and have managed to understand several of the above questions:
    1. Understood
    2. Understood, but would like to see more detail in Tutorial: Hotfix Data.
    3. Not Understood. Is 'Original Gearbox hotfix data' required as part of every patch/hotfix? Why are these included in the Community Patch?
    4. Not Understood. I still have not learned how to obj dump to disk.
    5. The answer to #4 will probably solve this one.
    6. Swapping. I have studied in some detail the Enemy Randomizer. In simple terms, it appears the author or authors appropriated 3 or 4 mixes, assigned various objects to them, tweaked probability and assigned these mixes to the maps. I am much impressed - this is quite inventive.

I have managed to run the Enemy Randomizer and, as advertised, it is quite the hog. Friends who join cannot see most of the enemies. I wish to do some swapping by script but have no need of random enemies. I know which ones I want to encounter and where I want to encounter them. Loading just the levels required and swapping just one map will be much easier on memory consumption and I feel it may just allow those who join to see the swapped enemies - without them also running the script.

I continue to study the randomizer code and to develop a simple script. I'll get there eventually. In the meantime, if someone could post just the minimum required to effect the swap of one special enemy for one other (special or a member of a mix) in a single map, it would be extremely helpful. Or any combination of special/mix members - it doesn't matter. Once I get my head around a simple example, I can figure out the rest.

**Comment by infamous27x** [comment]
Everyone needs to have Enemy Randomizer, enemies will show up as invisible, as they do not have the actual enemies loaded in on their client.

**Comment by Ravenjim** [comment]
Allrigth then for 
#3 If you DO NOT put the original GBX hotfix in the hotfix you make then it will be deleted. that is true for anything. if you keep anything that was in at the start and remove it then it will be deleted.

#4 There are two ways to get the dumps one is via filter tool which can obj dump outside of the game and is what everyone is using since its faster really. Or the classic way of going in game doing whatever you do on console and then going. C\User\NameOfUser\Documents\My Games\Borderlands The Pre-Sequel\WillowGame\Logs The latest Log file in there is the last time you were in game.

I cannot really asnwer #6 as i dont really understand what you need to know for it. If you want to see the names of enemies on a single map then do " getall AiPawnbalanceDefinition name"

**Comment by Ravenjim** [comment]
In case you use the Filter tool you dont need to add to a large hotfix file code after code for every little thing you want to do as the filter tool merges the hotfix codes itself so you dont have to recode the whole thing everytime.

**Comment by EntityDev** [comment]
@infamous27x 
Thank you.

@Ravenjim 
Thank you for your reply. I very much appreciate it.
No. 3 - I don't understand your answer. What will be deleted? The hotfix that I create? I have created several, have not included the original Gearbox hotfix, and executed them. They are not "deleted". They run just fine. It is unclear what you mean.

Also, you stated, " if you keep anything that was in at the start and remove it then it will be deleted". I apologize, but I've read and re-read this for a couple days now and I confess I am unable to understand what this means. Can you elaborate, please?

No. 4 - Thank you very much. I explored the features of the Filter Tool and was able to generate logs. They are formatted well enough for my use.

No. 6 - No worries, and thank you once again. The 'getall' class name you provided gives me what I need.

**Regarding swapping:**
I have succeeded in figuring out the minimum steps required to swap an enemy. I simply initialize the ActorArchetypeList with the number of members I require, then assign an enemy or a mix to each of the member positions.

**Further study:**
I have not figured out what the Den Modifications are doing in the Enemy Randomizer. I have also found that they do not seem to be necessary in order to effect an enemy swap. Can someone tell me what the purpose of these line are, please?

Thanks again.

**Comment by Ravenjim** [comment]
For N3 what i mean is that the ORIGINAL hotfix will be deleted if you do not include it. Similar to any other hotfix if you do not include them all into ONE BIG ARRAY.

As for the remove-delete. Lets say a weapon has an X fire rate right? And then you see that the barrel adds another Y Fire rate.

Lets say the barrel has the below buffs on it.
WeaponAttributes (yada yada... WeaponDamage K , FireIntereval Y , MagazineSize L )
Now . That Gun without that barrel has a fire rate of 1.0 and with the barrel it has 1.3 right?
If you do.
WeaponAttributes (yada yada... WeaponDamage K  , MagazineSize L )
It will drop back to 1.0 as the Y buff is no longer there THUS deleted..

If you have Discord you could DM me and id be able to help/respond Faster
@Jim Raven#3934

**Comment by EntityDev** [comment]
@Ravenjim Thank you. No worries about the time to respond - it's all good.

I think what you're saying is that if one executes a hotfix, then executes a second, the first hotfix will be disregarded. Whether this is true or not, I use the Filter Tool to combine or merge everything in the files I write already.

What I was really wondering is whether or not the section comprising the original Gearbox hotfix is necessary. Why is it included in the file?

The answer to this becomes less important as I progress. The scripts I'm writing are working fine without adding the original Gearbox fix. I am just curious why it's there.

**Comment by Ravenjim** [comment]
As i said. The original Hotfix is in the same array as in the Hotfix array we use. If you do not include it it will not function.

Since you Use the Filter tool that doesnt do anything for you BUT you still need it inside your load order.
